Enhanced SubjectEntity Query System

jhweir opened the pull request:
# Enhanced SubjectEntity Query System
This PR introduces a new, more powerful query system for SubjectEntity that enables complex queries with a fluent interface and real-time subscriptions to query results.

# Key Features
1. Unified Query Building: All instance queries now use a single Prolog query that fetches all properties, collections, and metadata in one go
2. Rich Query Constraints: Support for filtering by:
 - Properties and their values
 - Source
 - Collections
 - Author and timestamp metadata
 
3. Result Shaping: Control over what data is returned via:
 - Property selection
 - Collection selection
 - Ordering
 - Pagination (limit/offset)

4. Real-time Query Subscriptions: Subscribe to query results and receive updates when matching instances change
**Huge for a much more performant implementation of our hooks! - coming next**

5. Fluent Interface: Chain-able query building for better readability

# Example usage
```JavaScript
// Basic query with where clause
const recipes = await Recipe.findAll(perspective, { 
  where: { name: "Pancakes" } 
});

// Complex query with multiple constraints
const recipes = await Recipe.findAll(perspective, {
  where: { 
    name: "Pancakes",
    author: currentUser.did
  },
  properties: ["name", "ingredients"],
  collections: ["comments"],
  limit: 10,
  offset: 0
});

// Using the fluent query builder
const query = Recipe.query(perspective)
  .where({ name: "Pancakes" })
  .properties(["name", "ingredients"])
  .collections(["comments"])
  .limit(10);

// Run once
const results = await query.run();

// Or subscribe to updates
query.subscribeAndRun((recipes) => {
  console.log("Query results updated:", recipes);
});
```

# Implementation Details
The new query system works by:
1. Building a comprehensive Prolog query that includes all necessary constraints
2. Executing the query to get all matching instances and their data in a single round-trip
3. Converting the Prolog results into properly typed SubjectEntity instances
4. For subscriptions, watching for changes and re-running the query automatically

This replaces the previous approach of fetching instances and their data separately, resulting in better performance and enabling more complex queries.

jhweir committed:
Initial experiments adding query logic to SubjectEntity all function

lucksus committed:
Cache subject class name in SubjectEntity

jhweir committed:
Unused id prop removed from getData

jhweir committed:
propertyGettersQuery function set up, used in getData() & findAll(), & tested in prolog-and-literals.test

lucksus committed:
Resolve subject properties in client if needed

jhweir committed:
Resolve & class name fixes, prop name updates for clarity

jhweir committed:
mapPropertiesToInstance function set up

jhweir committed:
fmt

jhweir committed:
Timestamp & author included in getData query

jhweir committed:
collectionGetterQuery set up

jhweir committed:
SubjectClass defintion extracted from getter queries

jhweir committed:
Include collections in findAll

jhweir committed:
Include timestamp & author in findAll

jhweir committed:
Prolog findall predicate moved to getter functions

jhweir committed:
authorAndTimeStampQuery extracted

jhweir committed:
Common prolog queries stores as consts

jhweir committed:
Property name changes

jhweir committed:
Updated prolog-and-literals tests

lucksus commented on core/src/subject/SubjectEntity.ts:
```diff
@@ -8,26 +8,62 @@ export type QueryPartialEntity<T> = {
+   * Constructs a new subject.
+   * @param perspective - The perspective that the subject belongs to.
+   * @param baseExpression - The base expression of the subject.
+   * @param soruce - The source of the subject, the expression this instance is linked too.
```
```suggestion
   * @param source - The source of the subject, the expression this instance is linked too.
```

lucksus commented on core/src/subject/SubjectEntity.ts:
```diff
@@ -49,75 +85,169 @@ export class SubjectEntity {
-    Object.assign(this, data);
-    this.#baseExpression = tempId;
-    return this
+  private static async assignValuesToInstance(
```
Why have this be a static method when it takes instance as parameter? Could be `instance.assignFromPrologResults()` or so...

jhweir committed:
Properties query set up with tests

jhweir committed:
Collections query set up with tests

jhweir committed:
buildAuthorAndTimestampQuery function set up, code refactor

jhweir committed:
Formatting, notes, & Query type

jhweir committed:
Initial where query set up

jhweir committed:
Test findAll() works with a mix of query constraints

jhweir committed:
Where query works with timestamp & author

jhweir committed:
Where query works with not operations

jhweir committed:
Where query works with arrays of timestamps and authors

jhweir committed:
Convert number strings to numbers in literal_from_url predicate

jhweir committed:
Boolean URI set up & boolean conversion in literal_from_url predicate

jhweir committed:
New where query structure & operation testing

jhweir committed:
Where query reorganised, comments added

jhweir committed:
Old methods removed

jhweir committed:
Source prop added to query params

jhweir committed:
WhereOps type, ops formatting

lucksus commented on core/src/subject/SubjectEntity.ts:
```diff
@@ -8,26 +8,208 @@ export type QueryPartialEntity<T> = {
+  //     property(SubjectClass, PropertyName),
+  //     property_getter(SubjectClass, Base, PropertyName, PropertyURI),
+  //     (property_resolve(SubjectClass, PropertyName) -> Resolve = true ; Resolve = false),
+
```
something like:
```prolog
Resolve -> expression_data_url(PropertyValue, PropertyURI) ; PropertyValue = PropertyURI
```
with
```prolog
expression_data_url(ExpressionData, Uri) :- 
  language_name(Uri, "literal") -> (
    literal_from_url(Uri, Literal, Scheme),
    (
      (Scheme = "json" -> (json_property(Literal, "data", ExpressionData); ExpressionData = Literal))
      ;
      ExpressionData = Literal
    )
  )
```

lucksus commented on core/src/subject/SubjectEntity.ts:
```diff
@@ -8,26 +8,208 @@ export type QueryPartialEntity<T> = {
+  return (Object.entries(where) as [string, WhereCondition][])
+    .map(([key, value]) => {
+      const isSpecial = ["author", "timestamp"].includes(key);
+      const getter = `property_getter(SubjectClass, Base, "${key}", URI), literal_from_url(URI, V, _)`;
```
if the URI is not a literal, I guess literal_from_url won't match the property is omitted

lucksus replied:
should be handled generically with property query code above

lucksus commented on core/src/subject/SubjectEntity.ts:
```diff
@@ -8,26 +8,208 @@ export type QueryPartialEntity<T> = {
+      else return `${getter}, V = ${formatValue(key, value)}`;
+    })
+    .join(", ");
+}
```
nice!

lucksus commented on core/src/subject/SubjectEntity.ts:
```diff
@@ -49,75 +231,192 @@ export class SubjectEntity {
+      await Promise.all(
+        values.map(async ([name, value, resolve]) => {
+          // Resolve the value if necessary
+          const finalValue = resolve ? (await perspective.getExpression(value)).data : value;
```
So if we have the literal handling in Prolog as above, we will need to understand here that we don't have to resolve those values. Maybe we should have an internal `ResolveInternal` prolog variable above and only if it's not a literal language URI set `Resolve = true`...

lucksus commented on core/src/subject/SubjectEntity.ts:
```diff
@@ -49,75 +231,192 @@ export class SubjectEntity {
+    const requestedAttribtes = [...(query?.properties || []), ...(query?.collections || [])];
+    const allInstances = await Promise.all(
+      result[0].AllInstances.map(async ([Base, Properties, Collections, Timestamp, Author]) => {
+        const instance = new this(perspective, Base) as any;
```
While this `Base` *should* really be a base of `this` Subject Class, defensive programming would be to wrap all of this in a `try {} catch()`, log errors and filter out the elements that couldn't be instantiated...

lucksus commented on core/src/subject/SubjectEntity.ts:
```diff
@@ -49,75 +231,192 @@ export class SubjectEntity {
+      })
+    );
+
+    return allInstances;
```
👏 👏 👏 
I guess next step could be to extract the prolog result parsing into its own function, well basically all 3 steps (query creation, infer, result parsing), so that we can have a version that only runs once, and one that subscribes to the query which we want to use in the hooks.

lucksus commented on core/src/subject/SubjectEntity.ts:
```diff
@@ -49,75 +231,192 @@ export class SubjectEntity {
+        resolveLanguage = resolveLanguageResults[0].Language;
       }
 
       if (resolveLanguage) {
```
Hm, weird, so we do have the code that would create the literal URLs from values when resolve language is set. Should investigate if and/or when that is failing...

lucksus commented on tests/js/tests/prolog-and-literals.test.ts:
```diff
@@ -570,6 +570,22 @@ describe("Prolog + Literals", () => {
+                    // todo: allow boolean type and auto convert to URI
+                    @SubjectProperty({
+                        through: "recipe://boolean_test",
+                        writable: true,
```
```suggestion
                        writable: true,
                        resolveLanguage: "literal",
```

lucksus commented on tests/js/tests/prolog-and-literals.test.ts:
```diff
@@ -570,6 +570,22 @@ describe("Prolog + Literals", () => {
+                    // todo: allow number type and auto convert to URI
+                    @SubjectProperty({
+                        through: "recipe://number",
+                        writable: true,
```
```suggestion
                        writable: true,
                        resolveLanguage: "literal",
```

lucksus reviewed (commented):
amazing work so far!

jhweir committed:
Where prop default value fix

lucksus commented on tests/js/tests/prolog-and-literals.test.ts:
```diff
@@ -818,6 +802,476 @@ describe("Prolog + Literals", () => {
+                    const recipesWithNameOnly = await Recipe.findAll(perspective!, { properties: ["name"] });
+                    expect(recipesWithNameOnly[0].name).to.equal("recipe://test_name")
+                    expect(recipesWithNameOnly[0].booleanTest).to.be.undefined
+                    expect(recipesWithNameOnly[0].author).to.be.undefined
```
I'm not sure about the utility of constraining it to certain properties, but I could fiture a case where we have big properties (especially if at some point we would have automatic loading/instantiation as default).. but even if we would have that, I think author and timestamp should always be there.

lucksus committed:
In Prolog property resolution for literals

lucksus committed:
“where” constraint on content of resolved property

Extracted property getter query with literal resolving out into const and reused in buildWhereQuery()

jhweir committed:
sort_instances predciate set up

jhweir committed:
Ordering set up

jhweir committed:
Ordering tests

jhweir committed:
Limit & offset set up with tests

jhweir committed:
Less than or equal to & greater than or equal to set up with tests

jhweir committed:
formatValue updates

jhweir committed:
lt, lte, gt, & gte operation handling reorganised

jhweir committed:
Broken tests & logic commented out for now

jhweir committed:
sdna.rs fmt

jhweir committed:
Manual literal creation in tests removed

lucksus committed:
Make json_property/3 work with number and boolean values

lucksus committed:
Use PROPERTY_RESOLVE_QUERY again

lucksus committed:
Replace dangerous cut-operator if plain logic

! (cut) would prevent the findall from backtracking more property solutions, basically omitting all properties after one that was resolved through json_property

lucksus committed:
Leave “name” prop to not resolve

lucksus committed:
Make “name” resolved again, add unresolved “plain” and adjust assertions

lucksus committed:
Unflake test that uses timestamp by adding sleeps

lucksus committed:
Extract parts from findAll() into query building and result parsing functions

lucksus committed:
QueryBuilder with subscribeAndRun(callback)

lucksus committed:
Test and fix query subscriptions actually returning correct data

jhweir committed:
JSON parse subscription query results, AllInstancesResult type added

jhweir committed:
Try catch block added to subcription query JSON parsing

jhweir committed:
JSON parsing removed from integration tests

lucksus committed:
Update core/src/subject/SubjectEntity.ts

jhweir committed:
Author & Timestamp always included in queries (no longer optional)

jhweir committed:
Unused types removed

lucksus committed:
Remove unused and commented-out test case

jhweir committed:
Option to pass single Query object to the query fucntion set up

lucksus committed:
Test to ensure SubjectQueryBuilder uses the right class to match to

lucksus committed:
Fix assingValuesToInstance in case resolving URI failed

lucksus committed:
Fix static classNameByClass cache to actually work for sub classes

lucksus committed:
Fix test assertion of new SubjectQueryBuilder test

jhweir committed:
Formatting & unused type export removed

lucksus committed:
Fix SubjectQueryBuilder to use constructor of subclass for correct matching

lucksus committed:
Merge branch 'ad4m-record' of github.com:coasys/ad4m into ad4m-record

# Conflicts:
#	core/src/subject/SubjectEntity.ts
#	tests/js/tests/prolog-and-literals.test.ts

lucksus committed:
Increase timeout on timestamp based test to avoid flaky results

lucksus committed:
Turn PROPERTY_RESOLVE_QUERY into built-in predicate

lucksus committed:
Unflake, complex query builder test for single query object

lucksus committed:
Test: findAll() works with multiple property constraints in one where clause

lucksus committed:
Use inline class instead of Recipe in “query builder works with subscriptions” to unflake

lucksus committed:
Use inline class instead of Recipe in “findAll() works with where query between operations” to unflake

lucksus committed:
Merge branch 'dev' into ad4m-record

lucksus committed:
Changelog

jhweir committed:
Error handling in instancesFromPrologResult, export type Query, & formatting
